### PR TITLE
GH-73991: Add `pathlib.Path.move()`

### PR DESCRIPTION
**Pull request by barneygale**
Add a `Path.move()` method that moves a file or directory tree, and returns a new `Path` instance pointing to the target.

This method is similar to `shutil.move()`, except that it doesn't accept a *copy_function* argument, and it doesn't check whether the destination is an existing directory.

In pathlib's private ABCs, `PathBase.move()` uses the `copy()` and `delete()` methods to move files and directories. 

* Issue: gh-73991

----
📚 Documentation preview 📚: https://cpython-previews--122073.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/pathlib.rst
Comment (picnixz) [inline Doc/library/pathlib.rst]: What happens if target is the same as the source file? does it raise like for `mv` or is a no-op?

File: Lib/test/test_pathlib/test_pathlib_abc.py
Comment (picnixz) [inline Lib/test/test_pathlib/test_pathlib_abc.py:2138]: Add a test that moves a file to itself as well (and not just a directory).

File: Doc/library/pathlib.rst
Comment (barneygale) [reply Doc/library/pathlib.rst]: It raises - I've added a mention in the paragraph above.

File: Lib/test/test_pathlib/test_pathlib_abc.py
Comment (barneygale) [reply Lib/test/test_pathlib/test_pathlib_abc.py:2138]: Done!

Comment (picnixz) [reply Lib/test/test_pathlib/test_pathlib_abc.py:2138]: Thanks!

File: Lib/pathlib/_abc.py
Comment (picnixz) [inline Lib/pathlib/_abc.py]: Don't you need a return here? or something else? Otherwise, the flow continues normally.

File: Lib/pathlib/_abc.py
Comment (picnixz) [inline Lib/pathlib/_abc.py]: I didn't check, but should you set filename and filename2 for this? (it could be useful for callbacks).

File: Lib/pathlib/_local.py
Comment (picnixz) [inline Lib/pathlib/_local.py]: Ditto.

File: Doc/library/pathlib.rst
Comment (picnixz) [inline Doc/library/pathlib.rst]: I'm not sure about the last point. You say that `.move` works as `shutil.move` but for the latter:

> Recursively move a file or directory (src) to another location and return the destination.
>  If dst is an existing directory or a symlink to a directory, then **src is moved inside that directory**. The destination path in that directory must not already exist.

So... is `move` not allowed to move a file like `mv file dir` and get `dir/file` ?

File: Lib/test/test_pathlib/test_pathlib_abc.py
Comment (picnixz) [inline Lib/test/test_pathlib/test_pathlib_abc.py:2101]: For instance, here, I would have expected `base / 'dirB' / 'fileA'` to be created.

File: Doc/library/pathlib.rst
Comment (barneygale) [reply Doc/library/pathlib.rst]: > So... is `move` not allowed to move a file like `mv file dir` and get `dir/file` ?

That's correct, and there's a little bit of discussion around this on the issue: https://github.com/python/cpython/issues/73991#issuecomment-1901041013

If we detect existing destination directories and add a basename to the path, the real destination is unpredictable to users unless they already know whether the destination is a directory, or they bother to check the return value. In an interactive shell you probably already know whether the `mv` target is an existing directory, but that's less likely to be the case in an Python program that uses `Path.move()`.

(It's perhaps worth noting that only some `shutil` functions support existing directories as targets, e.g. `copy()` does but not `copyfile()`.)

To move a path into an existing directory, you could do:

```python
def move_into(source, target_dir):
    source.move(target_dir / source.name)
```

File: Lib/pathlib/_abc.py
Comment (barneygale) [reply Lib/pathlib/_abc.py]: Yes indeed, thanks.

**Comment by barneygale** [comment]
Marking this as a draft as it will need to be revised once #122368 and #122369 land.

**Comment by barneygale** [comment]
Should be ready to review again :-)

This isn't _quite_ the last PR in this series. I'll log another to add `copy_into()` and `move_into()`, and then post to discuss.python.org about perhaps eliminating some/all of the *ignore_errors*, *on_error* and *ignore* arguments to `delete()` and `copy()`.

**Comment by barneygale** [comment]
Sorry for flip-flopping. I've moved the `copy()` changes into their own PR (#122924) so I'll mark this as a draft once again (-:

**Comment by picnixz** [comment]
I've reviewed the other PR. Do you want me to review this one now or should I wait?

**Comment by barneygale** [comment]
> I've reviewed the other PR. Do you want me to review this one now or should I wait?

Sorry for the slow reply - I think it would be best to wait for the other PR to land before reviewing this.

**Comment by barneygale** [comment]
Hey @picnixz, this is ready to review again FYI. Thanks in advance :)

File: Doc/library/pathlib.rst
Comment (picnixz) [inline Doc/library/pathlib.rst:1619]: Just me thinking loud, but do you think we should have a boolean to *not* overwrite an existing target (in which case, a `FileExistsError` would be raised)?

File: Lib/test/test_pathlib/test_pathlib_abc.py
Comment (picnixz) [inline Lib/test/test_pathlib/test_pathlib_abc.py:2080]: Maybe `self.assertFalse(target.exists())` before `source.move`?

File: Lib/test/test_pathlib/test_pathlib_abc.py
Comment (picnixz) [inline Lib/test/test_pathlib/test_pathlib_abc.py:2130]: You should check existence/absence of source/target before calling `source.move` as well.

File: Lib/test/test_pathlib/test_pathlib_abc.py
Comment (picnixz) [inline Lib/test/test_pathlib/test_pathlib_abc.py:2181]: Some missing asserts about the existence of the source I would say (and to its existence prior to the call).

File: Misc/NEWS.d/next/Library/2024-07-21-02-00-46.gh-issue-73991.pLxdtJ.rst
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-07-21-02-00-46.gh-issue-73991.pLxdtJ.rst:1]:
```suggestion
Add :meth:`pathlib.Path.move`, which moves a file or a directory tree.
```

**Review by picnixz — COMMENTED** [review]
I think it would be good to also have assertions before calling `move()`. That way you really know how the state of the files change before/after the test (and that way, we are sure that we are in a correct test environment). I did not mark all tests that need this but any test that does `self.assertTrue(target.exists())` after moving should do `self.assertFalse(target.exists())` before moving (and if the target is to be overwritten, then it should check that it's content changed).

File: Doc/library/pathlib.rst
Comment (barneygale) [reply Doc/library/pathlib.rst:1619]: We could add a `clobber=True` argument, but I think that would be a separate PR.

**Comment by barneygale** [comment]
I don't understand why we'd make assertions *before* calling `move()` - we already know the state of the test directory (`self.base`) because it's configured to our specification in `setUp()`. If we applied this logic consistently we'd need to adjust almost every test case in `DummyPathTest` and `PathTest`!

File: Doc/library/pathlib.rst
Comment (picnixz) [reply Doc/library/pathlib.rst:1619]: I would have thought of `strict` (strict=False by default, strict=True would raise) or `replace=True`. But we can address this question later. Otherwise, it's just a three liner where users would do `if not os.exists(target)` if they want to avoid replacing files.

**Comment by picnixz** [comment]
> I don't understand why we'd make assertions before calling move() - we already know the state of the test directory (self.base) because it's configured to our specification in setUp().

I shouldn't make review at 3 AM... Yes, why didn't I think about it. So yes, it's fine (I'm marking my comments as resolved). Sorry Barney for the time loss :(

**Comment by barneygale** [comment]
> I shouldn't make review at 3 AM... Yes, why didn't I think about it. So yes, it's fine (I'm marking my comments as resolved). Sorry Barney for the time loss :(

Oh now, don't be silly! I really appreciate all your help with this PR and others, I apologise if I came off too strongly in my previous comment. I consider these sorts of discussions a great use of my time :)

File: Doc/library/pathlib.rst
Comment (barneygale) [reply Doc/library/pathlib.rst:1619]: ["Clobber"](https://en.wikipedia.org/wiki/Clobbering) has some currency already, e.g. in `mv --no-clobber`. But yeah, let's discuss later.

File: Lib/test/test_pathlib/test_pathlib_abc.py
Comment (picnixz) [inline Lib/test/test_pathlib/test_pathlib_abc.py:2100]: Did we actually cover the case "move file to a different empty directory"? AFAIU, this wouldn't raise an OSError (what we did test was "move file to a file in a different directory"). Namely, did we cover:

```py
source = base / 'fileA'
target = base / 'emptydir'
target.mkdir(exists_ok=False)
source.move(target)
self.assertFalse(source.exists())
# I think the target now becomes a file according to os.replace() docs.
self.assertFalse(target.is_dir())
```

**Comment by picnixz** [comment]
> I apologise if I came off too strongly in my previous comment. I consider these sorts of discussions a great use of my time :)

No your comment was definitely legitimate! I also appreciate your review and feedback on my `fnmatch` issues / PRs!

File: Lib/test/test_pathlib/test_pathlib_abc.py
Comment (barneygale) [reply Lib/test/test_pathlib/test_pathlib_abc.py:2100]: > Did we actually cover the case "move file to a different empty directory"? AFAIU, this wouldn't raise an OSError 

The `os.replace()` docs actually don't specify what happens when the destination is an existing empty directory. In practice is varies slightly by OS, e.g. `os.replace(dir, empty_dir)` succeeds on Windows but fails on POSIX. I've been careful in the `move()` docs not to specify this either, as we're at the mercy of whatever `replace()` does. As such I don't think it's worth testing because it's not part of the contract we want to offer users!

Comment (picnixz) [reply Lib/test/test_pathlib/test_pathlib_abc.py:2100]: Oh. I see. Since the docs say that `os.replace` is called, I think it's fine then to leave it as is.

**Review by picnixz — APPROVED** [review]
All good for me! I'll leave the discussion on *clubber* opened so that you can create a separate issue from the comment if you want.

**Comment by barneygale** [comment]
Amazing, thanks so much! I'll get back to your `fnmatch` stuff shortly!

**Pull request merged**
